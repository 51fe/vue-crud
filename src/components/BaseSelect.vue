<template>
  <el-select
    v-model="current"
    :multiple="multiple"
    :placeholder="placeholder"
    filterable
    clearable
    class="base-select"
    v-bind="$attrs"
    v-on="$listeners"
  >
    <el-option
      v-for="(item, index) in options"
      :key="index"
      :label="item.label"
      :value="item.value"
    />
  </el-select>
</template>

<script>
const toNumber = (str) => isNaN?.(str) ? str : Number(str)
export default {
  name: 'BaseSelect',
  props: {
    value: {
      type: [Number, String],
      default: undefined
    },
    options: {
      type: Array,
      default: () => []
    },
    multiple: {
      type: Boolean,
      default: false
    },
    placeholder: {
      type: String,
      default: '请选择'
    }
  },
  computed: {
    current: {
      get() {
        const value = this.value
        if (value && typeof value === 'string') {
          if (this.multiple) {
            return value.split(',').map(item => toNumber(item))
          }
          return toNumber(value)
        }
        return value
      },
      set(value) {
        if(this.multiple) {
          value = value.join(',')
        }
        this.$emit('input', value)
      }
    }
  }
}
</script>

<style>
.base-select {
  width: 100%;
}
</style>
