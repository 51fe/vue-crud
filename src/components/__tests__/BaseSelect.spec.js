import { fireEvent, render, screen } from '@testing-library/vue'
import BaseSelect from '../BaseSelect.vue'

const options = [{
  id: 1,
  brand: '移动'
}, {
  id: 2,
  brand: '联通'
}, {
  id: 3,
  brand: '电信'
}]
const defaultPlaceholder = '请选择'

const setUp = (overrides) => {
  return render(BaseSelect, {
    props: {
      options,
      labelKey: 'brand',
      valueKey: 'id',
      ...overrides
    }
  })
}

const getOptionByLabel = (label) => screen.getByRole('listitem', {
  hidden: true,
  name: (content, el) => el.textContent === label
})

test('renders correctly by default', () => {
  const { html } = setUp()
  screen.getByPlaceholderText(defaultPlaceholder)
  const list = screen.getAllByRole('listitem', { hidden: true })
  expect(list).toHaveLength(options.length)
  expect(html()).toMatchSnapshot()
})

test('changes props correctly', async () => {
  const placeholder = '请选择营运商'
  const value = 3
  setUp({
    placeholder,
    value
  })
  // change placeholder and current selected
  screen.getByPlaceholderText(placeholder)
  expect(getOptionByLabel('电信')).toHaveClass('selected')
})

test('emits input event correctly when selected', async () => {
  const { emitted } = setUp()
  //  trigger an update event by clicking the <option> element.
  await fireEvent.click(getOptionByLabel('电信'))
  expect(emitted().input[0][0]).toBe(3)
})

test('Passes and returns a string when select multiple items', async () => {
  const value = '1, 2'
  const { emitted } = setUp({
    multiple: true,
    value
  })
  // current selected 移动, 联通
  value.split(',').forEach(id => {
    const found = options.find(item => item.id === parseInt(id))
    if (found) {
      getOptionByLabel(found.brand)
    }
  })
  // select all
  await fireEvent.click(getOptionByLabel('电信'))
  expect(emitted().input[0][0]).toBe('1,2,3')
})
