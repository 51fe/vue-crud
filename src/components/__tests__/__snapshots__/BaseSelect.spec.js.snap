// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly by default 1`] = `
<div class="el-select base-select">
  <!---->
  <div class="el-input el-input--suffix">
    <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
    <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----><!----><!----><!----></span>
    <!----></span>
    <!---->
    <!---->
  </div>
  <transition-stub name="el-zoom-in-top">
    <div class="el-select-dropdown el-popper" style="display: none;">
      <div class="el-scrollbar" style="display: none;">
        <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
          <ul class="el-scrollbar__view el-select-dropdown__list">
            <!---->
            <li class="el-select-dropdown__item"><span>移动</span></li>
            <li class="el-select-dropdown__item"><span>联通</span></li>
            <li class="el-select-dropdown__item"><span>电信</span></li>
          </ul>
        </div>
        <div class="el-scrollbar__bar is-horizontal">
          <div class="el-scrollbar__thumb" style="width: 0px; transform: translateX(0%); webkit-transform: translateX(0%);"></div>
        </div>
        <div class="el-scrollbar__bar is-vertical">
          <div class="el-scrollbar__thumb" style="height: 0px; transform: translateY(0%); webkit-transform: translateY(0%);"></div>
        </div>
      </div>
      <p class="el-select-dropdown__empty">
        无数据
      </p>
    </div>
  </transition-stub>
</div>
`;
